### PR TITLE
Fix project lifecycle switching

### LINKED ISSUE #1501: Make project switching a consistent lifecycle operation
## Problem

Project switching is fragile because the code does not model the project lifecycle explicitly.

`GtProject::isOpen()` currently infers whether a project is open from whether it has children:

```cpp
return childCount<GtObject*>() != 0;
```

This conflates:

- membership in a session,
- project data being present in memory, and
- the project being the current/active project.

A visible consequence is the workaround in `GtProjectUI::duplicateProject()`:

```cpp
gtDataModel->newProject(newProject, true);
gtDataModel->saveProject(newProject);
gtDataModel->closeProject(newProject);
gtDataModel->closeProject(project);
gtDataModel->openProject(newProject);
gtApp->setCurrentProject(newProject);
```

A duplicated project already has cloned children, so it is considered “open” before it has actually gone through the open/activation lifecycle. At the same time, `openProject()` refuses to proceed while another current project exists. The duplicate is therefore saved, unloaded, and reloaded solely to make switching work.

There are related consistency problems:

- `newProject(project, true)` ignores the result of `openProject()` and can report success even when opening failed.
- `appendProjectData()` may reject empty data while `openProject()` still reports success.
- `closeProject()` cannot close a project without children.
- Loading data implicitly calls `setCurrentProject()`.
- Closing implicitly searches for another “open” project using the same child-count heuristic.
- `GtSession::setCurrentProject()` permits a closed project to become current.
- UI switching ignores the final close/open results.

## Expected behavior

Switching should be one consistent operation:

1. optionally save the current project,
2. close/unload the current project,
3. open/load the target project,
4. make the target current only after loading succeeds.

At most one project should be open/current, and failures should leave the application in a well-defined state.

## Proposed direction

Keep the existing public APIs where practical and repair their semantics:

- Preserve `GtProject::isOpen()`, but back it with explicit internal lifecycle state rather than child count.
- Let only the data model modify that state.
- Separate loading/unloading project data from setting the current project.
- Make `openProject()`, `closeProject()`, and `newProject(..., true)` propagate real failures.
- Allow clearing the current project explicitly.
- Add an additive `switchProject(GtProject*)` operation as the authoritative core transition.
- Keep save/discard confirmation in the GUI layer.
- Change duplication to produce a persisted, closed project that can be added to the session and opened normally.
- Remove the duplicate-project workaround after the lifecycle is reliable.

This should be achievable with little or no source API breakage. If explicit state requires storage in `GtProject`, prefer its private implementation/PIMPL to limit ABI impact.

## Acceptance criteria

- [ ] `isOpen()` no longer depends on child count.
- [ ] A project cannot be current unless it is successfully open.
- [ ] Closing the current project results in a defined null-current state before another project is opened.
- [ ] Switching A → B is performed through one checked operation.
- [ ] A failed load of B does not make B current or leave partial project data attached.
- [ ] `newProject(..., true)` reports an opening failure.
- [ ] Empty valid projects can be opened and closed, or are rejected with an accurately propagated error.
- [ ] Duplicating and activating a project no longer requires save/close/close/open/manual-set-current calls.
- [ ] Focused tests cover open, close, A → B switching, load failure, empty projects, and populated duplicates.
- [ ] Existing module-facing project APIs remain source compatible unless a documented breaking change is justified.

### PR DESCRIPTION
**Pull request by rainman110**
## Description

Model project open state explicitly and use one checked close/open transition when switching projects. This removes the duplicate-project save/close/reopen workaround and prevents closed projects from becoming current.

Fixes #1501.

## How Has This Been Tested?

- Built `GTlabUnitTest`.
- Ran all 691 tests: 690 passed, one existing test skipped.
- Manually verified project switching in GTlab.

## Checklist:

- [x] A test for the new functionality was added (if applicable).
- [x] All tests run without failure.
- [ ] The changelog has been extended, if this MR contains important changes from the users's point of view.
- [x] The new code complies with the GTlab's style guide.
- [x] New interface methods / functions are exported via `EXPORT`. Non-interface functions are NOT exported.
- [x] The number of code quality warnings is not increasing.

## Summary by Sourcery

Model project open/closed state explicitly and introduce a single, validated lifecycle transition when switching projects to prevent closed or invalid projects from becoming current.

New Features:
- Add explicit project switching API in the core data model that closes the current project and opens the target project in one operation.

Bug Fixes:
- Ensure projects outside the active session or with corrupted/empty data cannot be opened or set as current.
- Prevent closed projects from being set as the current project and from appearing open based solely on having children.
- Fix duplicate project handling so duplicates are persisted and returned as closed projects following the normal open path.
- Rollback newly added projects when opening them fails, avoiding inconsistent session state.

Enhancements:
- Change project open state tracking to use an internal flag on GtProject instead of relying on child count.
- Tighten appendProjectData to validate input and return success/failure, allowing callers to handle lifecycle errors and roll back changes.
- Refine GUI project switching flow to prompt for saving changes and report errors when switching fails.

Tests:
- Add core data model lifecycle tests covering project open/close/switch behavior, rollback on failed opens, and duplicate project persistence.
- Add session tests to verify closed projects cannot become current.
- Extend project tests to cover isOpen behavior independent of child objects.
- Add a unit test ensuring appendProjectData rejects empty project data.

### CODE REVIEW

**Comment by rainman110** [comment]
The remaining uncovered lines are the cleanup path after final persistence of a duplicated project fails:

```cpp
if (!saved)
{
    delete retval;
    return nullptr;
}
```

This path cannot be tested reliably with the current design. Creating the duplicate already writes its initial project files, so reaching this later failure requires the same filesystem location to become unwritable between initial generation and final persistence (for example through a disk-full condition or a timing-dependent permission change). Simulating that directly would make the test flaky and platform-dependent. A deterministic test would require an injectable project-persistence failure/test seam, which is not justified solely for covering these defensive lines.

File: src/core/gt_coredatamodel.cpp [hunk 1]
```diff
@@ -70,15 +70,15 @@ GtCoreDatamodel::setSession(GtSession* session)
...
 GtCoreDatamodel::appendProjectData(GtProject* project,
                                    QList<GtObject*> const& projectData)
 {
-    if (projectData.isEmpty())
+    if (!project || projectData.isEmpty())
     {
         gtWarning() << tr("could not open project") << QStringLiteral(" - ") <<
                     tr("project data corrupted!");
-        return;
+        return false;
```
[the remark this answers is not in the document]
Comment (rainman110) [reply src/core/gt_coredatamodel.cpp]: Addressed in 33abb784: null projects and empty/corrupted project data now use separate validation branches and warning messages.

File: tests/unittests/core/test_gt_coredatamodel.cpp [hunk 2]
```diff
@@ -105,3 +122,135 @@ TEST_F(TestGtCoreDataModel, makeUnqiueName_null_parent)
...
+    ASSERT_NE(current, nullptr);
+    ASSERT_TRUE(gtDataModel->newProject(current, true));
+
+    GtProject* rejected = createProject(QStringLiteral("Rejected Project"));
+    ASSERT_NE(rejected, nullptr);
+    EXPECT_FALSE(gtDataModel->newProject(rejected, true));
+    EXPECT_EQ(gtDataModel->findProject(rejected->objectName()), nullptr);
+    EXPECT_EQ(rejected->parent(), nullptr);
+    delete rejected;
+}
```
[the remark this answers is not in the document]
Comment (rainman110) [reply tests/unittests/core/test_gt_coredatamodel.cpp:228]: Addressed in 33abb784: the rollback test now verifies that the original project remains open and remains the current project after opening the rejected project fails.

**Comment by rainman110** [comment]
Addressed the review’s overall comments in 33abb784:

- Project open-state writes are now centralized through private `GtProject::markOpen()` / `markClosed()` transition methods, without changing the public API, object layout, or virtual table.
- `appendProjectData()` rollback now calls `acceptChangesRecursively()` after deleting the appended children, matching the cleanup performed by `closeProject()`.
- `openProject()` now rejects projects outside the active session before loading their data. This also removes the Windows test hang caused by loading and then rolling back a detached project hierarchy.

Validation: the focused lifecycle/externalization sequence passes, and the full local suite reports 695 passed with one existing skipped test.

**Comment by rainman110** [comment]
@jensschmeink Could you please test the known existing issues with the project open / switch / upgrade actions?

**Comment by rainman110** [comment]
@mariusalexander Could you please review the pr? I have more in the pipeline that requires this to merge first

**Review by mariusalexander — APPROVED** [review]
LGTM

File: src/gui/object_ui/gt_projectui.cpp [hunk 3]
```diff
@@ -582,12 +602,14 @@ GtProjectUI::saveProjectAs(GtObject* obj)
             return;
         }
 
-        gtDataModel->newProject(newProject, true);
-        gtDataModel->saveProject(newProject);
-        gtDataModel->closeProject(newProject);
-        gtDataModel->closeProject(project);
-        gtDataModel->openProject(newProject);
-        gtApp->setCurrentProject(newProject);
```
Comment (mariusalexander) [inline src/gui/object_ui/gt_projectui.cpp:590]: Was this the old implementation?! Opening and closing the projects x-times... yikes :D Good improvement

Comment (rainman110) [reply src/gui/object_ui/gt_projectui.cpp:590]: Yes! That was insane and it took me a while why this was necessary 🤦
